waiter/src/waiter/scheduler.clj

```diff
@@ -638,3 +642,145 @@
+            instances-requested-delta (- (:requested instance-counts')
+                                         (:requested instance-counts))
+            instance-scheduling-start-times'
+            (->> now
+                 (repeat instances-requested-delta)
+                 (into instance-scheduling-start-times))
+            ;; Process timers for instances that finished scheduling
+            unmatched-instance-count (- (count new-instance-ids)
+                                        (count instance-scheduling-start-times'))
+            [matched-start-times instance-scheduling-start-times'']
+            (split-at (count new-instance-ids) instance-scheduling-start-times')
+            ;; Process timers for starting instances
+            {started-instance-ids :started starting-instance-ids' :starting}
+            (->> starting-instance-ids
+                 (remove removed-instance-id?)
+                 (concat new-instance-ids)
+                 (group-by #(if (contains? instance-id->heathly-instance %) :started :starting)))]
+        ;; Warn about any new instances for which we didn't track the scheduling time
+        (when (pos? unmatched-instance-count)
+          (log/warn unmatched-instance-count
+                    "untracked scheduled instances discovered in instance launch metric loop for service"
+                    service-id))
+        ;; Report schedule time for now-known instances
+        (doseq [start-time matched-start-times]
+          (metrics/report-duration service-schedule-timer start-time now)
+          (metrics/report-duration waiter-schedule-timer start-time now))
+        ;; Report startup time for instances that are now healthy
+        (doseq [instance-id started-instance-ids]
+          (let [{:keys [started-at]} (instance-id->heathly-instance instance-id)]
+            (metrics/report-duration service-startup-timer started-at now)))
+        ;; tracker-state'
+        (assoc tracker-state
+               :instance-counts instance-counts'
+               :known-instance-ids known-instance-ids'
+               :instance-scheduling-start-times (vec instance-scheduling-start-times'')
+               :starting-instance-ids (vec starting-instance-ids'))))))
+
+(defn start-launch-metrics-maintainer
+  "go block to collect metrics on the instance launch overhead of waiter services.
+
+  Updated state for the router is read from `router-state-chan`."
+  [router-state-chan]
+  (let [exit-chan (async/chan 1)
+        query-chan (au/latest-chan)
+        update-state-timer (metrics/waiter-timer "state" "launch-metrics-maintainer" "update-state")
+        waiter-schedule-timer (metrics/waiter-timer "launch-overhead" "schedule-time")]
+    (async/go
+      (try
+        (loop [{:keys [known-service-ids service-id->launch-tracker] :as current-state}
+               {:known-service-ids #{}
+                :service-id->launch-tracker {}}]
```

Comment:
The initial state for this loop should be initialized based on the router state at the time this worker comes up. Using empty collections here would result in garbage stats when a router restarts in a non "clean slate" state. I have a patch for this, but I'm still testing it. I'll push it later today.